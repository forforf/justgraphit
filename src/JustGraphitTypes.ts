
// Not perfect, but ok for now. Maybe improve if typescript supports regexes
export type ISODatetimeType = `${number}-${number}-${number}T${number}:${number}:${number}.${number}Z`;

export type GraphName = string;

export type DateOrIso = Date | ISODatetimeType;

// individual data points
export type JustGraphitEntry = {
  number: number;
  datetime: ISODatetimeType;
  dt?: Date;
}

export type JustGraphitStore = Record<GraphName, JustGraphitEntry[]>

export type PassthroughFunction = (...args: unknown[]) => unknown;

export type RefSvg = SVGSVGElement | null;

export type D3Select = (ref: RefSvg) => PassthroughFunction;

// Functions that handle graphs by graphName (changing graphs, deleting graphs, etc)
export type GraphHandler = (graphName: GraphName) => void;

export type GraphEntryHandler = (graphName: GraphName, numberInput: string) => void;