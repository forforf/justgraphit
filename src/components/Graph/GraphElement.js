import React from 'react';
import IsEmpty from 'lodash.isempty';
import EmptyGraph from './EmptyGraph.js'
import SingleItemGraph from './SingleItemGraph.js'
import LineGraph from './LineGraph.js'


function GraphElement({graphName, graphData, height, width}) {
  // Early returns for legibility
  if (IsEmpty(graphData)) return <EmptyGraph />
  // we know graphData is not empty
  if (graphData.length === 1) return <SingleItemGraph
    graphName={graphName}
    itemNumber={graphData[0].number}
    itemDatetime={graphData[0].datetime}
  />
  return <LineGraph
    chartId="jgi-line-graph"
    graphName={graphName}
    graphData={graphData}
    height={height}
    width={width}
  />
}

export default GraphElement
