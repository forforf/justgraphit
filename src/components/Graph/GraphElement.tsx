import React from 'react';
import IsEmpty from 'lodash.isempty';
import EmptyGraph from './EmptyGraph';
import SingleItemGraph from './SingleItemGraph';
import LineGraph from './LineGraph';
import { JustGraphitEntry } from '../../JustGraphitTypes';

export type GraphElementProps = {
  graphData: JustGraphitEntry[];
  height: number;
  width: number;
};

function GraphElement({
  graphData,
  height,
  width,
}: GraphElementProps): JSX.Element {
  // Early returns for legibility
  if (IsEmpty(graphData)) return <EmptyGraph />;
  // we know graphData is not empty
  if (graphData.length === 1)
    return (
      <SingleItemGraph
        itemNumber={graphData[0].number}
        itemDatetime={graphData[0].datetime}
      />
    );
  return (
    <LineGraph
      chartId="jgi-line-graph"
      graphData={graphData}
      height={height}
      width={width}
    />
  );
}

export default GraphElement;
