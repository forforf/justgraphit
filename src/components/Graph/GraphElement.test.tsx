import React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow'
import GraphElement, {GraphElementProps} from './GraphElement';
import {ISODatetimeType, JustGraphitEntry} from "../../JustGraphitTypes";


const height = 10;
const width = 20;
let graphData: JustGraphitEntry[];
let graphElementProps: GraphElementProps;


describe('empty graph', () => {
  beforeEach(() => {
    graphData = [];
    graphElementProps = {graphData, height, width};
  });

  test('deep snapshot', () => {
    const component = renderer.create(
      <GraphElement {...graphElementProps}/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe('single item graph', () => {
  beforeEach(() => {
    graphData = [{number: 1, datetime: (new Date('2000-01-01T00:00:00')).toISOString() as ISODatetimeType}];
    graphElementProps = {graphData, height, width};
  });

  test('deep snapshot', () => {
    const component = renderer.create(
      <GraphElement {...graphElementProps}/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });


  test('shallow snapshot', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render( <GraphElement {...graphElementProps}/>);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot()
  });
})

describe('line graph', () => {
  beforeEach(() => {
    graphData = [
      {number: 1, datetime: (new Date('1999-12-31T23:59:59')).toISOString() as ISODatetimeType},
      {number: 2, datetime: (new Date('2000-01-02T00:00:00')).toISOString() as ISODatetimeType},
    ];
    graphElementProps = {graphData, height, width};
  });

  test('deep snapshot', () => {
    const component = renderer.create(
      <GraphElement {...graphElementProps}/>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });


  test('shallow snapshot', () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render( <GraphElement {...graphElementProps}/>);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot()
  });
})


