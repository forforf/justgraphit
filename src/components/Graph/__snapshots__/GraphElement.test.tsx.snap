// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty graph deep snapshot 1`] = `
<div>
  This is an EmptyGraph
</div>
`;

exports[`line graph deep snapshot 1`] = `
<svg
  className="line-graph"
  height={10}
  id="jgi-line-graph"
  width={20}
>
  <g
    className="axis"
    transform="translate(50,20)"
  />
  <g
    className="axis"
    transform="translate(50,-10)"
  />
  <g
    transform="translate(50,20)"
  >
    <path
      d="M0,-30L-50,0"
      fill="none"
      stroke="#000"
      strokeLinecap="round"
    />
  </g>
</svg>
`;

exports[`line graph shallow snapshot 1`] = `
<LineGraph
  chartId="jgi-line-graph"
  graphData={
    Array [
      Object {
        "datetime": "1999-12-31T23:59:59.000Z",
        "number": 1,
      },
      Object {
        "datetime": "2000-01-02T00:00:00.000Z",
        "number": 2,
      },
    ]
  }
  height={10}
  width={20}
/>
`;

exports[`single item graph deep snapshot 1`] = `
<div
  style={
    Object {
      "borderStyle": "solid",
      "marginLeft": "5em",
      "marginTop": "2em",
      "paddingBottom": "2em",
      "paddingTop": "2em",
    }
  }
>
  <div>
    Not enough to graph yet. Only a single lonely value of:
  </div>
  <h3>
    1
  </h3>
  22 years ago
</div>
`;

exports[`single item graph shallow snapshot 1`] = `
<SingleItemGraph
  itemDatetime="2000-01-01T00:00:00.000Z"
  itemNumber={1}
/>
`;
